### PR TITLE
fix: Construction of OIDC endpoint when rootPath has no trailing slash

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Fixes https://github.com/stackabletech/nifi-operator/issues/716

Deploying nifi into the e2e-sec demo:

### With `rootPath: /realms/demo`
0.0.0-dev: `https://85.215.194.243:30584/realms/.well-known/openid-configuration` (obviously broken) => Caused by: org.springframework.web.client.HttpClientErrorException$NotFound: 404 Not Found: "{"error":"Realm does not exist"}"
This PR: `https://85.215.194.243:30584/realms/demo/.well-known/openid-configuration` => Correct :heavy_check_mark: 

### With `rootPath: /realms/demo/`

0.0.0-dev: `https://85.215.194.243:30584/realms/demo/.well-known/openid-configuration` => Correct :heavy_check_mark: 
This PR: `https://85.215.194.243:30584/realms/demo/.well-known/openid-configuration` => Correct :heavy_check_mark: 

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -31,5 +32,5 @@ tracing = "0.1"
 url = { version = "2.5.2" }
 xml-rs = "0.8"
 
-# [patch."https://github.com/stackabletech/operator-rs.git"]
-# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
+[patch."https://github.com/stackabletech/operator-rs.git"]
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "fix/oidc-endpoint-calculation" }
```
Comment (sbernauer) [inline Cargo.toml:36]: Reminder to switch to released version

**Comment by sbernauer** [comment]
https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/nifi-operator-it-custom/23 passed :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -30,6 +30,7 @@ All notable changes to this project will be documented in this file.
 - Switch from `flow.xml.gz` to `flow.json.gz` to allow seamless upgrades to version 2.0 ([#675]).
 - Failing to parse one `NifiCluster`/`AuthenticationClass` should no longer cause the whole operator to stop functioning ([#662]).
 - NiFi will now use the JDK trust store when an OIDC provider uses WebPKI as CA ([#686], [#698]).
+- Fix OIDC endpoint calculation in case the `rootPath` does not have a trailing slash ([#718]).
```
Comment (Techassi) [inline CHANGELOG.md:33]: nit: We are not "calculating" the OIDC endpoint, rather we construct it. Please also update the PR title.

File: rust/operator-binary/src/security/oidc.rs [hunk 3]
```diff
@@ -25,15 +31,21 @@ pub enum Error {
...
-        STACKABLE_OIDC_ADMIN_PASSWORD_KEY
+        "found existing admin password secret {secret:?}, but the key {STACKABLE_OIDC_ADMIN_PASSWORD_KEY} is missing",
     ))]
-    OidcAdminPasswordKeyMissing { name: String, namespace: String },
+    OidcAdminPasswordKeyMissing { secret: ObjectRef<Secret> },
+
+    #[snafu(display("invalid OIDC well known URL"))]
+    InvalidOidcWellKnownUrl {
+        source: stackable_operator::commons::authentication::oidc::Error,
+    },
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:41]: suggestion: Make it clear, that this refers to the well-known OIDC configuration URL.

```suggestion
    #[snafu(display("invalid well-known OIDC configuration URL"))]
    InvalidWellKnownConfigUrl {
        source: stackable_operator::commons::authentication::oidc::Error,
    },
```

_Note: This also needs updating of the context selector below_

File: rust/operator-binary/src/security/oidc.rs [hunk 4]
```diff
@@ -25,15 +31,21 @@ pub enum Error {
...
 
     #[snafu(display(
-        "found existing secret [{}/{}], but key {} is missing",
-        name,
-        namespace,
-        STACKABLE_OIDC_ADMIN_PASSWORD_KEY
+        "found existing admin password secret {secret:?}, but the key {STACKABLE_OIDC_ADMIN_PASSWORD_KEY} is missing",
     ))]
-    OidcAdminPasswordKeyMissing { name: String, namespace: String },
+    OidcAdminPasswordKeyMissing { secret: ObjectRef<Secret> },
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:36]: note: Technically, we can drop the the `Oidc` prefix, as it is already implied by the error / the module.

```suggestion
    MissingAdminPasswordKey { secret: ObjectRef<Secret> },
```

_Note: This also needs updating of the context selector below_

File: rust/operator-binary/src/security/oidc.rs [hunk 5]
```diff
@@ -25,15 +31,21 @@ pub enum Error {
...
+
+    #[snafu(display("invalid OIDC well known URL"))]
+    InvalidOidcWellKnownUrl {
+        source: stackable_operator::commons::authentication::oidc::Error,
+    },
+
+    #[snafu(display("Nifi doesn't support skipping the OIDC TLS verification"))]
+    NoOidcTlsVerificationNotSupported {},
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:44]:
```suggestion
    UnsupportedTlsVerification {},
```

_Note: This also needs updating of the context selector below_

File: rust/operator-binary/src/security/oidc.rs [hunk 6]
```diff
@@ -25,15 +31,21 @@ pub enum Error {
...
 }
 
-/// Generate a secret containing the password for the admin user that can access the API. This admin user is the same as for SingleUser authentication.
+/// Generate a secret containing the password for the admin user that can access the API. This admin
+/// user is the same as for SingleUser authentication.
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:48]: nit: Add a newline.

```suggestion
/// Generate a secret containing the password for the admin user that can access the API.
///
/// This admin user is the same as for SingleUser authentication.
```

File: rust/operator-binary/src/security/oidc.rs [hunk 7]
```diff
@@ -92,3 +104,116 @@ pub(crate) async fn check_or_generate_oidc_admin_password(
 pub fn build_oidc_admin_password_secret_name(nifi: &NifiCluster) -> String {
     format!("{}-oidc-admin-password", nifi.name_any())
 }
+
+pub fn add_oidc_config(
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:108]: suggestion: Rename the function to `add_oidc_config_to_properties` to make it clearer to _what_ we are adding the OIDC config. Also, add doc comments to the function.

File: rust/operator-binary/src/security/oidc.rs [hunk 8]
```diff
@@ -92,3 +104,116 @@ pub(crate) async fn check_or_generate_oidc_admin_password(
...
+    provider: &oidc::AuthenticationProvider,
+    oidc: &ClientAuthenticationOptions,
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:110]: suggestion: Make it clearer what type this parameter holds.

```suggestion
    client_auth_options: &ClientAuthenticationOptions,
```

_Note: Needs adjustments below._

File: CHANGELOG.md (hunk 2 above)
Comment (sbernauer) [reply CHANGELOG.md]: Pls see 2684d86 and PR title

File: rust/operator-binary/src/security/oidc.rs (hunk 5 above)
Comment (sbernauer) [reply rust/operator-binary/src/security/oidc.rs:44]: I think the original name was more descriptive and would prefer to not lose the descriptiveness.
Maybe we can agree on something a bit more readable such as `SkippingOidcTlsVerificationNotSupported`, `SkippingOidcTlsVerificationUnsupported` or `UnsupportedOidcTlsVerificationSkipping` (although I prefer the first ^^)?

File: rust/operator-binary/src/security/oidc.rs (hunk 7 above)
Comment (sbernauer) [reply rust/operator-binary/src/security/oidc.rs]: Please have a look at d0352b9

File: rust/operator-binary/src/security/oidc.rs (hunk 5 above)
Comment (Techassi) [reply rust/operator-binary/src/security/oidc.rs:44]: I'm fine with `SkippingTlsVerificationNotSupported`.

File: rust/operator-binary/src/security/oidc.rs [hunk 9]
```diff
@@ -92,3 +105,120 @@ pub(crate) async fn check_or_generate_oidc_admin_password(
 pub fn build_oidc_admin_password_secret_name(nifi: &NifiCluster) -> String {
     format!("{}-oidc-admin-password", nifi.name_any())
 }
+
+/// Adds all the needed configuration properties that are needed to enable OIDC authentication.
```
Comment (Techassi) [inline rust/operator-binary/src/security/oidc.rs:109]:
```suggestion
/// Adds all the required configuration properties to enable OIDC authentication.
```

**Review by Techassi — APPROVED** [review]

**Comment by lfrancke** [comment]
Did this make it into 24.11?

**Comment by sbernauer** [comment]
Nope, but it will be part of 24.11.1. Not sure what label we should use here ^^

**Comment by lfrancke** [comment]
I think we should add a snippet to the 24.11.1 release notes for this (one entry for all of the affacted ones). Could you add a snippet here?

**Pull request merged**